p2p/conn/connection.go

```diff
@@ -645,9 +675,19 @@ FOR_LOOP:
 				break FOR_LOOP
 			}
 			if msgBytes != nil {
-				c.Logger.Debug("Received bytes", "chID", channelID, "msgBytes", msgBytes)
-				// NOTE: This means the reactor.Receive runs in the same thread as the p2p recv routine
-				c.onReceive(channelID, msgBytes)
+				msgLen := uint32(len(msgBytes))
+				bufferMaxSize := uint32(c.config.RecvMessageCapacity)
+				if msgLen > bufferMaxSize {
+					bufferMaxSize = msgLen
+				}
+				c.msgRecvSize += msgLen
+				c.Logger.Debug("Received bytes", "chID", channelID, "msgBytesLength", msgLen, "msgBytesBufferedLength", c.msgRecvSize, "pendingMsgsAhead", len(c.msgRecvQueue))
+				// Wait until our message fits in the queue
+				for c.msgRecvSize > bufferMaxSize {
+					c.msgRecvSize -= <-c.msgRecvAck
+				}
```

Comment:
`msgRecvSize` seems misnamed, something like `msgRecvBufferedByteCount` or `msgBytesBufferedLength` (matching the log message) would be more better IMO.